src/node_platform.cc

```diff
@@ -0,0 +1,201 @@
+  } else {
+    TracingController* controller = new TracingController();
+    tracing_controller_.reset(controller);
+  }
+  threads_.resize(thread_pool_size);
+  for (int i = 0; i < thread_pool_size; i++) {
+    uv_thread_create(&threads_[i], BackgroundRunner, &background_tasks_);
+  }
+}
+
+NodePlatform::~NodePlatform() {
+  background_tasks_.Stop();
+  for (size_t i = 0; i < threads_.size(); i++) {
+    uv_thread_join(&threads_[i]);
+  }
+  uv_close(reinterpret_cast<uv_handle_t*>(&flush_tasks_), DeleteHandle);
+}
+
+size_t NodePlatform::NumberOfAvailableBackgroundThreads() {
+  return threads_.size();
+}
+
+static void RunForegroundTask(uv_timer_t* handle) {
+  Task* task = static_cast<Task*>(handle->data);
+  task->Run();
+  delete task;
+  uv_close(reinterpret_cast<uv_handle_t*>(handle), DeleteHandle);
+}
+
+void NodePlatform::DrainBackgroundTasks() {
+  FlushForegroundTasksInternal();
+  background_tasks_.BlockingDrain();
+}
+
+void NodePlatform::FlushForegroundTasksInternal() {
+  std::pair<Task*, double>* delayed = NULL;
+  while ((delayed = foreground_delayed_tasks_.Pop())) {
+    uint64_t delay_millis =
+      static_cast<uint64_t>(delayed->second + 0.5) * 1000;
+    uv_timer_t* handle = new uv_timer_t();
+    handle->data = static_cast<void*>(delayed->first);
+    uv_timer_init(loop_, handle);
+    // Timers may not guarantee queue ordering of events with the same delay if
+    // the delay is non-zero. This should not be a problem in practice.
+    uv_timer_start(handle, RunForegroundTask, delay_millis, 0);
+    uv_unref(reinterpret_cast<uv_handle_t*>(handle));
+    delete delayed;
+  }
+  Task* task = NULL;
+  while ((task = foreground_tasks_.Pop())) {
+    task->Run();
+    delete task;
+  }
+}
+
+void NodePlatform::CallOnBackgroundThread(Task* task,
+                                          ExpectedRuntime expected_runtime) {
+  background_tasks_.Push(task);
+}
+
+void NodePlatform::CallOnForegroundThread(Isolate* isolate, Task* task) {
+  foreground_tasks_.Push(task);
+  uv_async_send(&flush_tasks_);
+}
+
+void NodePlatform::CallDelayedOnForegroundThread(Isolate* isolate,
+                                                 Task* task,
+                                                 double delay_in_seconds) {
+  std::pair<Task*, double>* pair =
+    new std::pair<Task*, double>(task, delay_in_seconds);
+  foreground_delayed_tasks_.Push(pair);
+  uv_async_send(&flush_tasks_);
+}
+
+bool NodePlatform::IdleTasksEnabled(Isolate* isolate) { return false; }
+
+double NodePlatform::MonotonicallyIncreasingTime() {
+  // Convert nanos to seconds.
+  return uv_hrtime() / 1e9;
+}
+
+TracingController* NodePlatform::GetTracingController() {
+  return tracing_controller_.get();
+}
+
+template <class T>
+TaskQueue<T>::TaskQueue()
+    : lock_(), tasks_available_(), tasks_drained_(),
+      outstanding_tasks_(0), stopped_(false), task_queue_() {
+  uv_mutex_init(&lock_);
+  uv_cond_init(&tasks_available_);
+  uv_cond_init(&tasks_drained_);
```

Comment:
`CHECK_EQ` return values here.  That said, this looks like it can be replaced with `node::Mutex` and `node::ConditionVariable` from src/node_mutex.h.